pennylane/devices/qubit/apply_operation.py

```diff
@@ -273,31 +295,45 @@ def apply_mid_measure(
     mid_measurements = execution_kwargs.get("mid_measurements", None)
     rng = execution_kwargs.get("rng", None)
     prng_key = execution_kwargs.get("prng_key", None)
-
     if is_state_batched:
         raise ValueError("MidMeasureMP cannot be applied to batched states.")
-    if not np.allclose(np.linalg.norm(state), 1.0):
-        mid_measurements[op] = -1
-        return np.zeros_like(state)
     wire = op.wires
-    sample = qml.devices.qubit.sampling.measure_with_samples(
-        [qml.sample(wires=wire)], state, Shots(1), rng=rng, prng_key=prng_key
-    )
-    sample = int(sample[0])
-    mid_measurements[op] = sample
-    if op.postselect is not None and sample != op.postselect:
-        mid_measurements[op] = -1
-        return np.zeros_like(state)
     axis = wire.toarray()[0]
     slices = [slice(None)] * qml.math.ndim(state)
-    slices[axis] = int(not sample)
-    state[tuple(slices)] = 0.0
-    state_norm = np.linalg.norm(state)
-    state = state / state_norm
+    slices[axis] = 0
+    prob0 = qml.math.norm(state[tuple(slices)]) ** 2
+    interface = qml.math.get_deep_interface(state)
+    if prng_key is not None:
+        # pylint: disable=import-outside-toplevel
+        import jax
+
+        _, key = jax.random.split(prng_key)
```

Comment:
You're right. We achieve a minimal amount of splitting with some global view of the code, but I wonder whether it wouldn't be better to take a bottom-up view for programmers with less knowledge about the code. This way it does not matter where the key comes from (say a new function by an external contributor), it remains reusable one level up in the stack. That said, I'm curious what you think but I don't really mind either way.

Reply:
I like your idea, we would also avoid meaningless splits this way as currently the key is split for every op in the tape (though it has no functional implications).

Reply:
You still need to split at the top level because we are not returning the first key. Otherwise we will always obtain the same key here. I'm rather thinking of future code that might call `apply_operation` for some reason and then sample a prob-dist. Then even if the developer doesn't split prior to sampling, the key is still unused. So it is safer to split at both levels albeit sometimes useless. If we want to keep splitting to a minimum tough, I would remove the splitting in `apply_operation` and maybe split in `get_final_state` depending on `isinstance`.